### PR TITLE
Convert to promise API

### PR DESCRIPTION
**Pull request by callumlocke**
see #56

### CODE REVIEW

File: cli.js [hunk 1]
```diff
@@ -1,9 +1,12 @@
 #!/usr/bin/env node
 'use strict';
+var loudRejection = require('loud-rejection');
 var meow = require('meow');
 var updateNotifier = require('update-notifier');
 var psi = require('./');
 
+loudRejection();
```
Comment (sindresorhus) [inline cli.js]: Not needed as it's already included in `meow` ;)

File: README.md [hunk 2]
```diff
@@ -34,7 +34,9 @@ psi.output('html5rocks.com', function (err) {
 
 ## API
 
-### psi(url, [options], callback)
+### psi(url, [options])
+
+Returns a promise that resolves with the response data from Google PageSpeed Insights.
```
Comment (sindresorhus) [inline README.md]: > Returns a promise for the response data from Google PageSpeed Insights.

File: index.js [hunk 3]
```diff
@@ -12,47 +13,36 @@ function handleOpts(url, opts) {
...
-      cb(err);
-      return;
+    if (!url) {
+      throw new Error('URL required');
     }
 
-    cb(null, response);
+    pagespeed(handleOpts(url, opts), function (err, response) {
+      if (err) {
+        err.noStack = true;
```
Comment (sindresorhus) [inline index.js]: You can drop the `noStack` thing. In hindsight it was a bad idea and it complicates the code.

File: index.js [hunk 4]
```diff
@@ -12,47 +13,36 @@ function handleOpts(url, opts) {
...
-    if (err) {
-      err.noStack = true;
-      cb(err);
-      return;
+    if (!url) {
+      throw new Error('URL required');
     }
 
-    cb(null, response);
+    pagespeed(handleOpts(url, opts), function (err, response) {
```
Comment (sindresorhus) [inline index.js]: You can use [pify](https://github.com/sindresorhus/pify) here and anywhere else where there are callback-style APIs.

File: cli.js [hunk 5]
```diff
@@ -30,15 +33,13 @@ if (!cli.input[0]) {
...
-
+psi.output(cli.input[0], cli.flags).then(function () {
   process.exit(0);
+}, function (err) {
+  if (err.noStack) {
+    console.error(err.message);
+    process.exit(1);
+  } else {
+    throw err;
+  }
```
Comment (sindresorhus) [inline cli.js]: Because of https://github.com/addyosmani/psi/pull/61/files#r43915908, you can remove this catch function.

**Comment by sindresorhus** [comment]
:+1: Lgtm when inline comments are fixed and merge conflict resolved :)

// @addyosmani

**Comment by callumlocke** [comment]
I think that's all done @sindresorhus.

@addyosmani let me know if you'd like me to squash the whole PR into one commit.

File: lib/output.js [hunk 6]
```diff
@@ -60,32 +60,35 @@ function getReporter(format) {
...
+      throw err;
+    }
+
+    if (parameters.download) {
+      var download = new Download()
+        .get(optimizedResoruceURL)
+        .dest('.')
+        .rename('./optimized.zip');
+
+      return pify(download.run.bind(download), Promise)();
```
Comment (sindresorhus) [inline lib/output.js]: You need to require `pify` in this file.

Comment (sindresorhus) [inline lib/output.js]: Fixed it for you https://github.com/addyosmani/psi/commit/ee69ee848f4ddfe52176cadc5e3b7e022d1068ba.

**Comment by sindresorhus** [comment]
Looks good. Thanks @callumlocke :)

**Comment by addyosmani** [comment]
Thanks for helping land this both <3

**Comment by callumlocke** [comment]
@sindresorhus that's cool, how did you make it so GitHub credits two users in a single commit?

image

**Comment by sindresorhus** [comment]
```
$ git commit --amend --author "name" -C HEAD
```

**Comment by callumlocke** [comment]
neat

**Pull request closed without merging**